ccl/changefeedccl: TestChangefeedNemeses failed

ccl/changefeedccl.TestChangefeedNemeses [failed](https://mesolite.cluster.engflow.com/invocations/default/83530cd4-285d-4502-97c7-5a8c87a1e632?testReportRun=3&testReportShard=8&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [ac68bcaf4a547b9d24127a6da8ee01df7753fea1](https://github.com/cockroachdb/cockroach/commits/ac68bcaf4a547b9d24127a6da8ee01df7753fea1):


```
=== RUN   TestChangefeedNemeses
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestChangefeedNemeses1522505897
    test_log_scope.go:82: use -show-logs to present logs inline
    nemeses_test.go:73: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedNemeses1522505897
--- FAIL: TestChangefeedNemeses (58.48s)
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true
--- FAIL: TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true (58.48s)
=== RUN   TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/cloudstorage
    helpers_test.go:1394: making server as system tenant
    helpers_test.go:1470: making cloudstorage feed factory
    nemeses_test.go:33: random seed: 6925798597782338345
    nemeses_test.go:40: using legacy schema changer
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
    nemeses_test.go:52: topic foo partition : saw new row timestamp 1759037222418376122.0000000000 after 1759037224077408801.0000000000 was seen
--- FAIL: TestChangefeedNemeses/nemeses_options==fp_validator=true,sql_smith=true/cloudstorage (58.35s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=8</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedNemeses.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -108,9 +108,10 @@ type changeAggregator struct {
 	eventConsumer eventConsumer
 
 	flushFrequency time.Duration // how often high watermark can be checkpointed.
+	lastSpanFlush  time.Time     // last time expensive, span based checkpoint was written.
 
 	// frontierFlushLimiter is a rate limiter for flushing the span frontier
-	// to the coordinator.
+	// to the coordinator because of frontier advancement.
 	frontierFlushLimiter *saveRateLimiter
 
 	// frontier keeps track of resolved timestamps for spans along with schema change
@@ -480,6 +481,9 @@ func (ca *changeAggregator) Start(ctx context.Context) {
 
 	// Init heartbeat timer.
 	ca.lastPush = timeutil.Now()
+
+	// Generate expensive checkpoint only after we ran for a while.
+	ca.lastSpanFlush = timeutil.Now()
 }
 
 func (ca *changeAggregator) startKVFeed(
@@ -932,9 +936,14 @@ func (ca *changeAggregator) noteResolvedSpan(resolved jobspb.ResolvedSpan) error
 		ca.sliMetrics.setResolved(ca.sliMetricsID, ca.frontier.Frontier())
 	}
 
+	if ca.knobs.ShouldFlushFrontier != nil && ca.knobs.ShouldFlushFrontier(resolved) {
+		return ca.flushFrontier(ctx)
+	}
+
 	forceFlush := resolved.BoundaryType != jobspb.ResolvedSpan_NONE
 
-	checkpointFrontier := (advanced && forceFlush) || ca.frontierFlushLimiter.canSave(ctx)
+	// TODO(#155015): Re-enable periodic frontier flushing.
+	checkpointFrontier := advanced && (forceFlush || ca.frontierFlushLimiter.canSave(ctx))
 
 	if checkpointFrontier {
 		now := crtime.NowMono()
@@ -942,6 +951,19 @@ func (ca *changeAggregator) noteResolvedSpan(resolved jobspb.ResolvedSpan) error
 			return err
 		}
 		ca.frontierFlushLimiter.doneSave(now.Elapsed())
+		return nil
+	}
+
+	// At a lower frequency, we checkpoint specific spans in the job progress
+	// either in backfills or if the highwater mark is excessively lagging behind.
+	checkpointSpans := (ca.frontier.InBackfill(resolved) || ca.frontier.HasLaggingSpans(sv)) &&
+		canCheckpointSpans(sv, ca.lastSpanFlush)
+
+	if checkpointSpans {
+		defer func() {
+			ca.lastSpanFlush = timeutil.Now()
+		}()
+		return ca.flushFrontier(ctx)
 	}
 
 	return nil
@@ -1761,7 +1783,7 @@ func (cf *changeFrontier) forwardFrontier(resolved jobspb.ResolvedSpan) error {
 		// The feed's checkpoint is tracked in a map which is used to inform the
 		// checkpoint_progress metric which will return the lowest timestamp across
 		// all feeds in the scope.
-		cf.sliMetrics.setCheckpoint(cf.sliMetricsID, newResolved)
+		cf.sliMetrics.setCheckpoint(cf.sliMetricsID, cf.frontier.Frontier())
 
 		return cf.maybeEmitResolved(cf.Ctx(), newResolved)
 	}

--- a/pkg/ccl/changefeedccl/changefeed_progress_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_progress_test.go
@@ -234,6 +234,10 @@ func TestChangefeedProgressSkewMetrics(t *testing.T) {
 					}
 					return false, nil
 				}
+
+				knobs.ShouldFlushFrontier = func(rs jobspb.ResolvedSpan) bool {
+					return true
+				}
 			}
 
 			// Create changefeed for both tables with no initial scan.
@@ -269,7 +273,8 @@ WITH no_initial_scan, min_checkpoint_frequency='1s', resolved, metrics_label='%s
 					}
 					if !perTableTracking {
 						if tableSkew != 0 {
-							return errors.Newf("expected table skew to be 0, got %d", tableSkew)
+							// TODO(#155083): Return an error here.
+							return nil
 						}
 						return nil
 					}

--- a/pkg/ccl/changefeedccl/testing_knobs.go
+++ b/pkg/ccl/changefeedccl/testing_knobs.go
@@ -71,6 +71,10 @@ type TestingKnobs struct {
 		partitions []sql.SpanPartition, draining []roachpb.NodeID,
 	) ([]sql.SpanPartition, error)
 
+	// ShouldFlushFrontier returns true if the change aggregator should flush
+	// its frontier after processing a resolved span.
+	ShouldFlushFrontier func(rs jobspb.ResolvedSpan) bool
+
 	// ShouldCheckpointToJobRecord returns true if change frontier should checkpoint itself
 	// to the job record.
 	ShouldCheckpointToJobRecord func(hw hlc.Timestamp) bool

--- a/pkg/cmd/roachtest/tests/cdc.go
+++ b/pkg/cmd/roachtest/tests/cdc.go
@@ -2146,6 +2146,9 @@ CONFIGURE ZONE USING
 		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
 			runCDCInitialScanRollingRestart(ctx, t, c, cdcFrontierPersistence)
 		},
+		// TODO(#155015): Unskip this test.
+		Skip: "frontier persistence will not happen during an initial-scan only changefeed " +
+			"without periodic aggregator frontier flushes",
 	})
 	r.Add(registry.TestSpec{
 		Name:             "cdc/fine-grained-checkpointing",

